### PR TITLE
feat!: var output estimation optimization

### PR DESCRIPTION
**Pull request by segfault-magnet**
closes: #1380 
The number of variable outputs is now determined by filling the dry-run tx with as much variable outputs as the node will allow (`consensus_params.tx_params.max_ouputs` - `whatever other outputs are already present`) and counting the number of `TransferOut` receipts made by the dry run. 

I've checked with the core devs, ideally we'd be looking at the `tx` outputs themselves but since the dry run endpoint doesn't return a `tx` this is the next best thing since only `TRO` instructions require a variable output attow. 

### Breaking changes

- `.append_variable_outputs(n)` is replaced by `.with_variable_output_policy(VariableOutputPolicy::Exactly(n))`.

- `.estimate_tx_dependencies(max_attempts)` has been replaced with `.determine_missing_contracts(max_attempts)` and will only add the missing contract inputs and no longer estimate variable outputs. To estimate variable outputs you should now add `.with_variable_output_policy(VariableOutputPolicy::EstimateMinimum)`

### Checklist
- [x] I have linked to any relevant issues.
- [x] I have updated the documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary labels.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Review by digorithm — CHANGES_REQUESTED** [review]
I spent quite some time debating whether the new API names are good or intuitive. After a while, they started growing on me, hah. I'm curious to hear everyone's opinions on these two new names. 

Great job on the actual solution; it's a very clever trick! 

I just left a small, doc-related suggestion. Other than that, lgtm.

File: packages/fuels-programs/src/script_calls.rs [hunk 1]
```diff
@@ -124,6 +119,14 @@ where
         self
     }
 
+    pub fn with_variable_output_policy(
```
[the remark this answers is not in the document]
Comment (segfault-magnet) [reply packages/fuels-programs/src/script_calls.rs:129]: [added ](https://github.com/FuelLabs/fuels-rs/pull/1393/commits/77a04a730a256a03b3292b60b5926f14ad487efa)

**Review by Br1ght0ne — DISMISSED** [review]

**Review by hal3e — DISMISSED** [review]
Awesome work! I really love the refactoring especially `ScriptDryRunner`! Thanks!

**Comment by hal3e** [comment]
Regarding the naming: `with_variable_output_policy` is good. However, `determine_missing_contracts` could be misleading because we will `determine` them and add them to the outputs. So maybe something like `fix_missing_contracts` or `determine_and_update_missing_contracts` ...

**Comment by digorithm** [comment]
> Regarding the naming: `with_variable_output_policy` is good. However, `determine_missing_contracts` could be misleading because we will `determine` them and add them to the outputs. So maybe something like `fix_missing_contracts` or `determine_and_update_missing_contracts` ...

Yeah, I was (and maybe still am) on the fence about `determine_missing_contracts`. Maybe just `update_missing_contracts`? `update` indicates it will mutate the inner stuff and also does some magic underneath.

**Comment by segfault-magnet** [comment]
Thing is previously it was called "estimate_tx_dependencies" that also didn't have anything claiming mutability. 

I share the concern but have no satisfactory name that isn't a kilometer long.

maybe `discover_missing_contract_inputs` ?

File: packages/fuels-core/src/types/transaction_builders/script_dry_runner.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        {
+            tx.inputs_mut().push(fake_input);
+
+            // Add an empty `Witness` for the `coin_signed` we just added
+            tx.witnesses_mut().push(Witness::default());
+            tx.set_witness_limit(tx.witness_limit() + WITNESS_STATIC_SIZE as u64);
+        }
+    }
+
+    fn needs_fake_base_input(
```
Comment (MujkicA) [inline packages/fuels-core/src/types/transaction_builders/script_dry_runner.rs:83]: There exists a tx validity condition that requires spendable inputs to be present:
https://github.com/FuelLabs/fuel-vm/blob/898789cc58ce549a92550d680df96fa0d0480002/fuel-tx/src/transaction/validity/error.rs#L15

I didn't come across anything that indicates that there needs to be base asset inputs. Running the tests with the base asset id check removed, seems to work fine.

Comment (segfault-magnet) [reply packages/fuels-core/src/types/transaction_builders/script_dry_runner.rs:83]: This is inherited. @hal3e do you remember the original reason why we've added the base asset req?

Comment (segfault-magnet) [reply packages/fuels-core/src/types/transaction_builders/script_dry_runner.rs]: @MujkicA I've tested it and it seems you're right, dry runs are passing without any base assets in inputs.

Something must have changed in the meantime @hal3e . 

In any case I've removed the requirement of the asset being of the base id.

**Review by Br1ght0ne — APPROVED** [review]

**Review by MujkicA — APPROVED** [review]

**Review by Salka1988 — APPROVED** [review]

**Review by hal3e — APPROVED** [review]

**Pull request merged**